parcels/interaction/interactionkernelsoa.py

```diff
@@ -0,0 +1,179 @@
+from parcels.tools.statuscodes import StateCode, OperationCode, ErrorCode
+from parcels.tools.statuscodes import recovery_map as recovery_base_map
+from parcels.tools.loggers import logger
+
+
+__all__ = ['InteractionKernelSOA']
+
+
+class InteractionKernelSOA(BaseInteractionKernel):
+    """InteractionKernel object that encapsulates auto-generated code.
+    """
+
+    def __init__(self, fieldset, ptype, pyfunc=None, funcname=None,
+                 funccode=None, py_ast=None, funcvars=None, c_include="",
+                 delete_cfiles=True):
+        super().__init__(fieldset=fieldset, ptype=ptype, pyfunc=pyfunc,
+                         funcname=funcname, funccode=funccode, py_ast=py_ast,
+                         funcvars=funcvars, c_include=c_include,
+                         delete_cfiles=delete_cfiles)
+
+        for func in self._pyfunc:
+            self.check_fieldsets_in_kernels(func)
+
+        numkernelargs = self.check_kernel_signature_on_version()
+
+        assert numkernelargs[0] == 5 and \
+            numkernelargs.count(numkernelargs[0]) == len(numkernelargs), \
+            'Interactionkernels take exactly 5 arguments: particle, fieldset, time, neighbors, mutator'
+
+        # At this time, JIT mode is not supported for InteractionKernels,
+        # so there is no need for any further "processing" of pyfunc's.
+
+    def execute_jit(self, pset, endtime, dt):
+        raise NotImplementedError
+
+    def __del__(self):
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.)
+        super().__del__()
+
+    def __add__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return self.merge(kernel, InteractionKernelSOA)
+
+    def __radd__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return kernel.merge(self, InteractionKernelSOA)
+
+    def execute_python(self, pset, endtime, dt):
+        """Performs the core update loop via Python"""
+        if self.fieldset is not None:
+            for f in self.fieldset.get_fields():
+                if type(f) in [VectorField, NestedField, SummedField]:
+                    continue
+                f.data = np.array(f.data)
+
+        for pyfunc in self._pyfunc:
+            pset.compute_neighbor_tree(endtime, dt)
+            active_idx = pset._active_particle_idx
+
+            mutator = defaultdict(lambda: [])
+            for particle_idx in active_idx:
+                p = pset[particle_idx]
+                # Don't use particles that are not started.
+                if (endtime-p.time)/dt <= -1e-7:
```

Comment:
so, it looks overly simplistic, so having a test with interaction kernel that does forward and backward simulation would be a good idea, to verify it that is indeed working. (haven't seen such a test existing, but I'll have a look at it again).

Apart from that, this part looks good. It's more the modifications that were required to the main loop that make me doubt the approach. But this section: yeah, well, if that works with backward- and forward simulations, and with dt's tested in the millisecond range (10^-3 seconds) over simulation times of a year overall (which is in the 10^6 seconds range, I think), then this seems to be solid enough to green-light it.

Reply:
I have added a small test for backward time integration. Though, I think simulating a year with ms dt's might be a bit though, since at ms resolution, simulation time is only slightly faster than real time...

Reply:
@qubixes I acknowledge and agree with your statement in full. Further, I agree it's probably not tenable at the moment to do that test. 

It is still though that, if that feature or change cannot be tested before more, then it is advisable to not apply or affect the change to other code branches that are tested and working on that issue. Otherwise, it's not 'defensive programming' but rather waiting for first follow-up comment on why those long-running small-dt simulations now suddenly break. Would be sad then to revert this whole issue to resolve a fix.

Reply:
@CKehl I understand your point. I think it's unlikely that it changes the behavior of the current code, but it definitely would be very sad if it did. Since it wouldn't involve much more than copying some pieces of code back and forth, I am willing/fine with reverting it to your suggestion for the merge, and creating a new branch after the merge that deduplicates the code (i.e. reverts it back to the current state of this branch). That gives you time to run the long time tests. I guess the following have to be tested?:

- Performance regressions with no interaction
- Time progression regressions with no interaction (small dt, large runtime, outputdt?)
- Time progression works well/the same with interaction (small dt, large runtime, outputdt?)

I'm willing to help with the tests, but it is probably easier/better if you do them?

Reply:
see new comment on the benchmmark results.